tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -413,40 +436,105 @@ private void warnIfCompleted(String feature) {
-                this.original = original;
-                this.translator = translator;
-                this.data = data;
-            }
+        TraceRestoringCloseableSpanWithMetadata(@Nullable Trace original, TagTranslator<? super T> translator, T data) {
+            this.original = original;
+            this.translator = translator;
+            this.data = data;
+        }
 
-            @Override
-            public void close() {
-                Tracer.fastCompleteSpan(translator, data);
-                Trace originalTrace = original;
-                if (originalTrace != null) {
-                    Tracer.setTrace(originalTrace);
-                }
+        @Override
+        public void close() {
+            Tracer.fastCompleteSpan(translator, data);
+            Trace originalTrace = original;
+            if (originalTrace != null) {
+                Tracer.setTrace(originalTrace);
             }
         }
     }
 
+    private static final class SampledUnnamedDetachedSpan implements DetachedSpan {
+
+        private final String traceId;
+        private final Optional<String> requestId;
+        private final OpenSpan openSpan;
+
+        SampledUnnamedDetachedSpan(String traceId, Optional<String> requestId, OpenSpan openSpan) {
+            this.traceId = traceId;
+            this.requestId = requestId;
+            this.openSpan = openSpan;
+        }
+
+        @Override
+        @MustBeClosed
+        public <T> CloseableSpan childSpan(
+                String operationName, TagTranslator<? super T> translator, T data, SpanType type) {
+            Trace maybeCurrentTrace = currentTrace.get();
+            setTrace(Trace.of(true, traceId, requestId));
+            Tracer.fastStartSpan(operationName, openSpan.getSpanId(), type);
+            return TraceRestoringCloseableSpanWithMetadata.of(maybeCurrentTrace, translator, data);
+        }
+
+        @Override
+        public DetachedSpan childDetachedSpan(String operation, SpanType type) {
+            return new SampledDetachedSpan(operation, type, traceId, requestId, Optional.of(openSpan.getSpanId()));
+        }
+
+        @Override
+        @MustBeClosed
+        public CloseableSpan attach() {
+            Trace maybeCurrentTrace = currentTrace.get();
+            Trace newTrace = Trace.of(true, traceId, requestId);
+            // Push the DetachedSpan OpenSpan to provide the correct parent information
+            // to child spans created within the context of this attach.
+            // It is VITAL that this span is never completed, it exists only for attribution.
+            newTrace.push(openSpan);
+            setTrace(newTrace);
+            // Do not complete the synthetic root span, it simply prevents nested spans from removing trace state, and
+            // allows
+            return maybeCurrentTrace == null ? REMOVE_TRACE : () -> Tracer.setTrace(maybeCurrentTrace);
```

Comment:
These methods are duplicated with the sampled implementation. Once I'm happier with the API I'll revisit this.